### PR TITLE
Introduce a ProvisionerStrategy to provision a node without delay

### PR DESCRIPTION
**Pull request by v1v**
### Description

Apparently the below properties should help to speed up the node provisioning:

```
System.setProperty('hudson.model.LoadStatistics.clock', '5000')
System.setProperty('hudson.model.LoadStatistics.decay', '0.5')
System.setProperty('hudson.agents.NodeProvisioner.MARGIN', '100')
System.setProperty('hudson.agents.NodeProvisioner.MARGIN0', '1.0')
```

but I'm afraid they don't work as expected. Even when using the `oneShot` flag.

### Tasks
- [x] Add changes suggested in the code review.
- [x] Integration Tests.

### Implementation

This particular change is similar to the done for other Cloud providers:
- https://github.com/jenkinsci/ec2-plugin/pull/294
- https://github.com/jenkinsci/amazon-ecs-plugin/pull/144
- https://github.com/jenkinsci/kubernetes-plugin/pull/598

In addition, it does include the fix [JENKINS-60055](https://issues.jenkins-ci.org/browse/JENKINS-60055)

For instance:

- running in parallel 60 stages to say `hey!` it takes 30 minutes or so

The very first stage to get a provisioned worker takes 3~4 minutes when there is a peak of 60 requests for workers:

image

The last stage waiting for a provisioned worker takes 23 minutes:

image

<details><summary>Expand to view the started provisioning log output</summary>
<p>

```
docker logs jenkins-lint 2>&1  | grep "Started provisioning" | sed 's#.*apply: ##g'

Started provisioning test-linux-kqprf7 from gce-test with 1 executors. Remaining excess workload: 0
Started provisioning test-linux-iit98i from gce-test with 1 executors. Remaining excess workload: 3.965
Started provisioning test-linux-0l6hp3 from gce-test with 1 executors. Remaining excess workload: 2.965
Started provisioning test-linux-4pdqlg from gce-test with 1 executors. Remaining excess workload: 1.965
Started provisioning test-linux-565sin from gce-test with 1 executors. Remaining excess workload: 0.965
Started provisioning test-linux-f7pt2n from gce-test with 1 executors. Remaining excess workload: -0.035
Started provisioning test-linux-ywtvjr from gce-test with 1 executors. Remaining excess workload: 4.168
Started provisioning test-linux-xd8had from gce-test with 1 executors. Remaining excess workload: 3.168
Started provisioning test-linux-5zuvf1 from gce-test with 1 executors. Remaining excess workload: 2.168
Started provisioning test-linux-s73pzb from gce-test with 1 executors. Remaining excess workload: 1.168
Started provisioning test-linux-xvvqap from gce-test with 1 executors. Remaining excess workload: 0.168
Started provisioning test-linux-9k8bdd from gce-test with 1 executors. Remaining excess workload: -0.832
Started provisioning test-linux-yvdiay from gce-test with 1 executors. Remaining excess workload: 5.806
Started provisioning test-linux-lzqd0s from gce-test with 1 executors. Remaining excess workload: 4.806
Started provisioning test-linux-o79gvf from gce-test with 1 executors. Remaining excess workload: 3.806
Started provisioning test-linux-gmbohi from gce-test with 1 executors. Remaining excess workload: 2.806
Started provisioning test-linux-g68jvv from gce-test with 1 executors. Remaining excess workload: 1.806
Started provisioning test-linux-d12ml0 from gce-test with 1 executors. Remaining excess workload: 0.806
Started provisioning test-linux-8mch6u from gce-test with 1 executors. Remaining excess workload: -0.194
Started provisioning test-linux-qx450a from gce-test with 1 executors. Remaining excess workload: 1.326
Started provisioning test-linux-2ahezy from gce-test with 1 executors. Remaining excess workload: 0.326
Started provisioning test-linux-xodzi1 from gce-test with 1 executors. Remaining excess workload: -0.674
Started provisioning test-linux-ohqlgc from gce-test with 1 executors. Remaining excess workload: 0.193
Started provisioning test-linux-2819jr from gce-test with 1 executors. Remaining excess workload: -0.807
Started provisioning test-linux-g7jn4k from gce-test with 1 executors. Remaining excess workload: -0.526
Started provisioning test-linux-vgcxqt from gce-test with 1 executors. Remaining excess workload: -0.474
Started provisioning test-linux-3dlcub from gce-test with 1 executors. Remaining excess workload: -0.626
Started provisioning test-linux-hzguiw from gce-test with 1 executors. Remaining excess workload: -0.367
Started provisioning test-linux-p0e8mm from gce-test with 1 executors. Remaining excess workload: -0.538
Started provisioning test-linux-nvyc3b from gce-test with 1 executors. Remaining excess workload: -0.85
Started provisioning test-linux-zkvttj from gce-test with 1 executors. Remaining excess workload: -0.216
Started provisioning test-linux-lw98nh from gce-test with 1 executors. Remaining excess workload: 8.282
Started provisioning test-linux-w80cfk from gce-test with 1 executors. Remaining excess workload: 7.282
Started provisioning test-linux-sjnxau from gce-test with 1 executors. Remaining excess workload: 6.282
Started provisioning test-linux-w9p8na from gce-test with 1 executors. Remaining excess workload: 5.282
Started provisioning test-linux-7d73c3 from gce-test with 1 executors. Remaining excess workload: 4.282
Started provisioning test-linux-b9t25u from gce-test with 1 executors. Remaining excess workload: 3.282
Started provisioning test-linux-5fu5bu from gce-test with 1 executors. Remaining excess workload: 2.282
Started provisioning test-linux-1vz92o from gce-test with 1 executors. Remaining excess workload: 1.282
Started provisioning test-linux-x509w1 from gce-test with 1 executors. Remaining excess workload: 0.282
Started provisioning test-linux-9rt8fl from gce-test with 1 executors. Remaining excess workload: 0.254
Started provisioning test-linux-lhrrp1 from gce-test with 1 executors. Remaining excess workload: 0.029
Started provisioning test-linux-jxu8pw from gce-test with 1 executors. Remaining excess workload: -0.271
Started provisioning test-linux-m044ir from gce-test with 1 executors. Remaining excess workload: 0.156
Started provisioning test-linux-ux8rsj from gce-test with 1 executors. Remaining excess workload: -0.844
Started provisioning test-linux-mnjkjl from gce-test with 1 executors. Remaining excess workload: 0.241
Started provisioning test-linux-gsxiht from gce-test with 1 executors. Remaining excess workload: -0.759
Started provisioning test-linux-r8xmbl from gce-test with 1 executors. Remaining excess workload: -0.083
Started provisioning test-linux-uxcyxh from gce-test with 1 executors. Remaining excess workload: -0.079
Started provisioning test-linux-ocswor from gce-test with 1 executors. Remaining excess workload: -0.271
Started provisioning test-linux-ypeu0s from gce-test with 1 executors. Remaining excess workload: -0.444
Started provisioning test-linux-d930pm from gce-test with 1 executors. Remaining excess workload: -0.8
Started provisioning test-linux-b74nms from gce-test with 1 executors. Remaining excess workload: -0.499
Started provisioning test-linux-q68j51 from gce-test with 1 executors. Remaining excess workload: -0.822
Started provisioning test-linux-1jnqrs from gce-test with 1 executors. Remaining excess workload: -0.477
Started provisioning test-linux-qno651 from gce-test with 1 executors. Remaining excess workload: -0.616
Started provisioning test-linux-a7dcp4 from gce-test with 1 executors. Remaining excess workload: -0.716
Started provisioning test-linux-nr6d11 from gce-test with 1 executors. Remaining excess workload: -0.781
Started provisioning test-linux-5gxtwh from gce-test with 1 executors. Remaining excess workload: -0.836
Started provisioning test-linux-aids9l from gce-test with 1 executors. Remaining excess workload: 0
```

</p>
</details>

The provisioning time seems to be lineal, so running the same `pipeline` for 10, 20, 30, 40, 50 and 60 parallel stages the overall time seems stable:

image

### Actions
- [x] Run the same test cases for this particular PR and compare it with.

The below overall execution team is based on the same `pipeline` for 10, 20, 30, 40, 60 and 100 parallel stages:

image

It seems it does cope with more concurrent workers although there is some kind of degradation. _NOTE_: the build with 30 workers took a bit longer as there were timeouts and I had to reset the timeout to be from 6 minutes to 10 minutes.

### UI

image

### References

Blocked by https://github.com/jenkinsci/google-compute-engine-plugin/issues/184

### Classloader

It seems the classloader with the remoting causes some issues with some threads (see [mailing-list](https://groups.google.com/forum/?nomobile=true#!topic/jenkinsci-dev/9SDhMTfW_yk)). For instance, I did add a particular random sleep before running the connection and there was some improvement in the overall time, see the below screenshot regarding the execution time for 60 nodes was faster than with 50 nodes.

image

## Related issues

- Blocked by https://github.com/jenkinsci/google-compute-engine-plugin/issues/189

### CODE REVIEW

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      Collections.shuffle(jenkinsClouds);
+      for (Cloud cloud : jenkinsClouds) {
+        int workloadToProvision = currentDemand - availableCapacity;
+        if (!(cloud instanceof ComputeEngineCloud)) continue;
+        if (!cloud.canProvision(label)) continue;
+        ComputeEngineCloud gcp = (ComputeEngineCloud) cloud;
+        if (!gcp.isNoDelayProvisioning()) continue;
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+          if (cl.canProvision(cloud, strategyState.getLabel(), workloadToProvision) != null) {
+            continue;
```
Comment (craigdbarber) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: This continue applies to the inner loop. What then is the intended effect of the loop?

**Comment by craigdbarber** [comment]
Thanks @v1v for this contribution! A few notes inline. Also, please be sure to run mvn verify locally and re-commit with the format changes :)

File: docs/Home.md [hunk 2]
```diff
@@ -110,3 +110,10 @@ Instance configurations have many options that were not listed above. A few of t
   access from metadata. For more info, review the service account documentation.
 
 
+# No delay provisioning
+
+By default Jenkins do estimate load to avoid over-provisioning of cloud nodes.
```
Comment (craigdbarber) [inline docs/Home.md]: s/do estimate/estimates

File: docs/Home.md [hunk 3]
```diff
@@ -110,3 +110,10 @@ Instance configurations have many options that were not listed above. A few of t
...
+This plugin will use its own provisioning strategy by default, with this strategy, a new node is created on GCP as soon as NodeProvisioner detects need for more agents.
+In worse scenarios, this will results in some extra nodes provisioned on Kubernetes, which will be shortly terminated.
```
Comment (craigdbarber) [inline docs/Home.md]: s/worse/worst case

Comment (craigdbarber) [reply docs/Home.md]: The strategy only applies to GCE based clouds if I'm not mistaken, so this wouldn't have an effect on k8s clouds.

File: docs/Home.md [hunk 4]
```diff
@@ -110,3 +110,10 @@ Instance configurations have many options that were not listed above. A few of t
...
+
+If you want to turn off this Strategy you can set SystemProperty `com.google.jenkins.plugins.computeengine.disableNoDelayProvisioning=true`
```
Comment (craigdbarber) [inline docs/Home.md]: Some users might not be as familiar with setting system properties directly, perhaps we should also mention the UI cloud configuration option.

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+import jenkins.model.Jenkins;
+
+/**
+ * Implementation of {@link NodeProvisioner.Strategy} which will provision a new node immediately as
+ * a task enter the queue.
+ */
+@Extension(ordinal = 100)
+public class NoDelayProvisionerStrategy extends NodeProvisioner.Strategy {
+
```
Comment (craigdbarber) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: Remove extra '\n'

File: src/main/resources/com/google/jenkins/plugins/computeengine/ComputeEngineCloud/help-noDelayProvisioning.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+By default Jenkins do estimate load to avoid over-provisioning of cloud nodes.
```
Comment (craigdbarber) [inline src/main/resources/com/google/jenkins/plugins/computeengine/ComputeEngineCloud/help-noDelayProvisioning.html]: s/do estimate/estimates

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
+package com.google.jenkins.plugins.computeengine;
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: Please add a license header.

**Comment by craigdbarber** [comment]
Would also like to see an int-test added which specifically tests the new strategy.

File: src/main/resources/com/google/jenkins/plugins/computeengine/ComputeEngineCloud/help-noDelayProvisioning.html (hunk 6 above)
Comment (stephenashank) [inline src/main/resources/com/google/jenkins/plugins/computeengine/ComputeEngineCloud/help-noDelayProvisioning.html]: Please add a license header.

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  private static final Logger LOGGER = Logger.getLogger(NoDelayProvisionerStrategy.class.getName());
+
+  private static final boolean DISABLE_NODELAY_PROVISING =
+      Boolean.valueOf(
+          System.getProperty(
+              "com.google.jenkins.plugins.computeengine.disableNoDelayProvisioning"));
+
+  @Override
+  public NodeProvisioner.StrategyDecision apply(NodeProvisioner.StrategyState strategyState) {
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: Please add a javadoc comment, even if it's just `{@InheritDoc}`

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  private static void fireOnStarted(
+      final Cloud cloud,
+      final Label label,
+      final Collection<NodeProvisioner.PlannedNode> plannedNodes) {
+    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+      try {
+        cl.onStarted(cloud, label, plannedNodes);
+      } catch (Error e) {
```
Comment (stephenashank) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: Could this be further specified or is this as much as would be known about the expected exceptions?

Comment (v1v) [reply src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: This particular method is a copy from the Jenkins NodeProvisioner class:
- https://github.com/jenkinsci/jenkins/blob/450d87491df16598409d26511bcad02978eaa8bb/core/src/main/java/hudson/slaves/NodeProvisioner.java#L903-L916

That particular error design was done in https://github.com/jenkinsci/jenkins/pull/2591

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java (hunk 1 above)
Comment (v1v) [reply src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]: That's caused by [JENKINS-60055](https://issues.jenkins-ci.org/browse/JENKINS-60055) and that's the default behaviour in the NodeProvisioner class too:
- https://github.com/jenkinsci/jenkins/blob/450d87491df16598409d26511bcad02978eaa8bb/core/src/main/java/hudson/slaves/NodeProvisioner.java#L721-L726

File: src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+public class NoDelayProvisionerStrategy extends NodeProvisioner.Strategy {
+  private static final Logger LOGGER = Logger.getLogger(NoDelayProvisionerStrategy.class.getName());
+
+  private static final boolean DISABLE_NODELAY_PROVISING =
+      Boolean.valueOf(
+          System.getProperty(
+              "com.google.jenkins.plugins.computeengine.disableNoDelayProvisioning"));
+
+  @Override
+  public NodeProvisioner.StrategyDecision apply(NodeProvisioner.StrategyState strategyState) {
```
Comment (v1v) [inline src/main/java/com/google/jenkins/plugins/computeengine/NoDelayProvisionerStrategy.java]:
```suggestion
  /** {@inheritDoc} */
  public NodeProvisioner.StrategyDecision apply(NodeProvisioner.StrategyState strategyState) {
```

**Comment by v1v** [comment]
@craigdbarber , I added some ITs with https://github.com/jenkinsci/google-compute-engine-plugin/pull/183/commits/5712a68f5b729b0b9be65b28c6422eca54508c78 but I'm afraid neither I cannot test them locally nor I know the status in the https://jenkins.cloudgraphite-cicd.net service, any ideas how I can debug what's going on? Thanks

**Comment by craigdbarber** [comment]
Hi @v1v I've granted you read access to the Jenkins server. You should be able to drill into the build/test logs now. Let me know if you encounter any issues. Thanks.

**Comment by viglesiasce** [comment]
@v1v were you able to access [the logs](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fgoogle-compute-engine-plugin/detail/PR-183/11/pipeline/)?

Here is the error:

```
[2020-04-27T21:01:42.342Z] [INFO] -------------------------------------------------------------
[2020-04-27T21:01:42.342Z] [ERROR] COMPILATION ERROR : 
[2020-04-27T21:01:42.342Z] [INFO] -------------------------------------------------------------
[2020-04-27T21:01:42.342Z] [ERROR] /C:/Jenkins/workspace/gle-compute-engine-plugin_PR-183/src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java:[90,41] unreported exception java.lang.InterruptedException; must be caught or declared to be thrown
[2020-04-27T21:01:42.342Z] [ERROR] /C:/Jenkins/workspace/gle-compute-engine-plugin_PR-183/src/test/java/com/google/jenkins/plugins/computeengine/integration/ConfigAsCodeTestIT.java:[91,43] unreported exception java.io.IOException; must be caught or declared to be thrown
```

**Comment by evandbrown** [comment]
LGTM if we can get that test passing. When that's done we'll merge and I'll cut a release.

File: pom.xml [hunk 11]
```diff
@@ -98,6 +98,27 @@
         <type>pom</type>
         <scope>import</scope>
       </dependency>
+      <!-- Require upper bound dependencies -->
```
Comment (v1v) [inline pom.xml:101]: This particular changes fixes the issues when using the default configuration in the ci.jenkins.io:
- https://github.com/jenkinsci/google-compute-engine-plugin/blob/develop/Jenkinsfile#L1

Errors:

```
14:01:51  [WARNING] Rule 4: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
14:01:51  Failed while enforcing RequireUpperBoundDeps. The error(s) are [
14:01:51  Require upper bound dependencies error for org.slf4j:jcl-over-slf4j:1.7.25 paths to dependency are:
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.slf4j:jcl-over-slf4j:1.7.25
14:01:51  and
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.jenkins-ci.main:jenkins-core:2.222.3
14:01:51      +-org.slf4j:jcl-over-slf4j:1.7.25 (managed) <-- org.slf4j:jcl-over-slf4j:1.7.26
14:01:51  , 
14:01:51  Require upper bound dependencies error for org.slf4j:log4j-over-slf4j:1.7.25 paths to dependency are:
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.slf4j:log4j-over-slf4j:1.7.25
14:01:51  and
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.jenkins-ci.main:jenkins-core:2.222.3
14:01:51      +-org.slf4j:log4j-over-slf4j:1.7.25 (managed) <-- org.slf4j:log4j-over-slf4j:1.7.26
14:01:51  , 
14:01:51  Require upper bound dependencies error for org.slf4j:slf4j-jdk14:1.7.25 paths to dependency are:
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.slf4j:slf4j-jdk14:1.7.25
14:01:51  and
14:01:51  +-org.jenkins-ci.plugins:google-compute-engine:4.2.1-SNAPSHOT
14:01:51    +-org.jenkins-ci.main:jenkins-war:2.222.3
14:01:51      +-org.slf4j:slf4j-jdk14:1.7.25 (managed) <-- org.slf4j:slf4j-jdk14:1.7.26
14:01:51  ]
```

**Comment by v1v** [comment]
I think we are good to go now! :) I did miss to run `mvn test-compile` for validating those compilation errors :(

Please, if you merge it please feel free to squash those commits

Thanks for your help

**Comment by evandbrown** [comment]
Thank you @v1v! Next I'll confirm tests are fine in `develop`, then manually verify by installing the HPI, then we'll release.

**Comment by dmaicher** [comment]
@v1v thanks for implementing and @evandbrown and everyone else involved thanks for testing and publishing it :+1: 

Using this new feature since yesterday without any issues and finally there is no weird 3min delay anymore until a new node is provisioned :tada:

**Pull request merged**